staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go

```diff
@@ -0,0 +1,336 @@
+	// Or nil if this validates a root object
+	parent *ratchetingValueValidator
+
+	// Path from parent to this validator or nil if this is a root validator
+	keyInParent interface{}
+
+	currentInvocation *ratchetInvocationOptions
+}
+
+// Context to track validation of a specific new value during a call to `Validate`
+type ratchetInvocationOptions struct {
+	thunk *ratchetingValueValidator
+
+	// Value being validated
+	value interface{}
+
+	// Cached comparison result of DeepEqual
+	comparisonResult *bool
+
+	// Cached map representation of a map-type list, or nil if not map-type list
+	mapList common.MapList
+
+	// Children spawned by a call to `Validate` on this object
+	children map[interface{}]*ratchetInvocationOptions
+}
+
+func (r *ratchetInvocationOptions) IsNoopLazy() (res bool) {
+	if r.comparisonResult != nil {
+		return *r.comparisonResult
+	}
+
+	defer func() {
+		r.comparisonResult = &res
+	}()
+
+	if r.value == nil && r.thunk.oldValue == nil {
+		return true
+	} else if r.value == nil || r.thunk.oldValue == nil {
+		return false
+	}
+
+	oldAsArray, oldIsArray := r.thunk.oldValue.([]interface{})
+	newAsArray, newIsArray := r.value.([]interface{})
+
+	if oldIsArray != newIsArray {
+		return false
+	} else if oldIsArray {
+		if len(oldAsArray) != len(newAsArray) {
+			return false
+		}
+
+		// Correctly considers map-type lists due to fact that index here
+		// is only used for numbering. The correlation is stored in the
+		// childInvocation itself
+		for i := range newAsArray {
+			// Query for child
+			childInvocation, ok := r.children[i]
+			if !ok {
+				return false
+			} else if !childInvocation.IsNoopLazy() {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	oldAsMap, oldIsMap := r.thunk.oldValue.(map[string]interface{})
+	newAsMap, newIsMap := r.value.(map[string]interface{})
+
+	if oldIsMap != newIsMap {
+		return false
+	} else if oldIsMap {
+		if len(oldAsMap) != len(newAsMap) {
+			return false
+		}
+
+		for k := range oldAsMap {
+			// Query for child
+			childInvocation, ok := r.children[k]
+			if !ok {
+				// Either key in old not in new, or validator short
+				// circuited
+				_, newAlsoHasKey := newAsMap[k]
+				if newAlsoHasKey {
+					// Child invocation is missing due to short circuit
+					return reflect.DeepEqual(r.thunk.oldValue, r.value)
+				}
+
+				// Child invocation is missing due to key change
+				return false
+			} else if !childInvocation.IsNoopLazy() {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	return reflect.DeepEqual(r.thunk.oldValue, r.value)
+}
+
+func newRatchetingValueValidator(oldValue interface{}, args schemaArgs, parent *ratchetingValueValidator, keyInParent interface{}) *ratchetingValueValidator {
+	return &ratchetingValueValidator{
+		oldValue:    oldValue,
+		schemaArgs:  args,
+		parent:      parent,
+		keyInParent: keyInParent,
+	}
+}
+
+func (r *ratchetingValueValidator) validateOption(value interface{}) validate.Option {
```

Comment:
`validateOption` is verb+object. Confusing name. `getValidateOptions` or `getValidateOptionsFor`.

Reply:
renamed to `getValidateOptionFor`